vm/systemSmartContracts/auction.go

```diff
@@ -1165,7 +1212,158 @@ func (s *stakingAuctionSC) claim(args *vmcommon.ContractCallInput) vmcommon.Retu
+	if err != nil {
+		s.eei.AddReturnMessage("cannot save registration data: error " + err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (s *stakingAuctionSC) computeMaxValueThatCanBeUnStaked(registrationData *AuctionDataV2) *big.Int {
+	maxVal := big.NewInt(0).Sub(registrationData.TotalStakeValue, registrationData.LockedStake)
+	maxVal.Sub(maxVal, registrationData.TotalUnstaked)
+
+	return maxVal
+}
+
+func (s *stakingAuctionSC) unBondTokens(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !s.flagStakingV2.IsSet() {
+		//unStakeTokens function will become available after enabling staking v2
+		s.eei.AddReturnMessage("invalid method to call")
+		return vmcommon.UserError
+	}
+
+	if len(args.Arguments) != 0 {
+		s.eei.AddReturnMessage("should have not specified any arguments")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		s.eei.AddReturnMessage(vm.TransactionValueMustBeZero)
+		return vmcommon.UserError
+	}
+	registrationData, err := s.getOrCreateRegistrationData(args.CallerAddr)
+	if err != nil {
+		s.eei.AddReturnMessage("cannot get registration data: error " + err.Error())
+		return vmcommon.UserError
+	}
+	if len(registrationData.RewardAddress) == 0 {
+		s.eei.AddReturnMessage("key is not registered, unStakeTokens is not possible")
+		return vmcommon.UserError
+	}
+	err = s.eei.UseGas(s.gasCost.MetaChainSystemSCsCost.Claim)
+	if err != nil {
+		s.eei.AddReturnMessage(vm.InsufficientGasLimit)
+		return vmcommon.OutOfGas
+	}
+
+	totalUnBond, index := s.computeUnBondTokens(registrationData)
+	if totalUnBond.Cmp(zero) == 0 {
+		s.eei.AddReturnMessage("no tokens that can be unbond at this time")
+		return vmcommon.Ok
+	}
+
+	registrationData.UnstakedInfo = registrationData.UnstakedInfo[index:]
+
+	registrationData.TotalUnstaked.Sub(registrationData.TotalUnstaked, totalUnBond)
+	if registrationData.TotalUnstaked.Cmp(zero) < 0 {
+		s.eei.AddReturnMessage("contract error on unBondTokens function, total unstaked < 0")
+		return vmcommon.UserError
+	}
+
+	err = s.eei.Transfer(args.CallerAddr, args.RecipientAddr, totalUnBond, nil, 0)
+	if err != nil {
+		s.eei.AddReturnMessage("transfer error on unBond function")
+		return vmcommon.UserError
+	}
+
+	err = s.saveRegistrationData(args.CallerAddr, registrationData)
+	if err != nil {
+		s.eei.AddReturnMessage("cannot save registration data: error " + err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (s *stakingAuctionSC) computeUnBondTokens(registrationData *AuctionDataV2) (*big.Int, int) {
+	var index int
+	var unstakedValue *UnstakedValue
+	currentNonce := s.eei.BlockChainHook().CurrentNonce()
+	totalUnBond := big.NewInt(0)
+	for index, unstakedValue = range registrationData.UnstakedInfo {
+		canUnbond := currentNonce-unstakedValue.UnstakedNonce >= s.unBondPeriod
+		if !canUnbond {
+			break
+		}
+
+		totalUnBond.Add(totalUnBond, unstakedValue.UnstakedValue)
+	}
+
+	return totalUnBond, index
```

Comment:
Should be checked through an unit test the situation when all the unStaked value could be unBonded and in this case which would be the value of the index returned. If the index will be the index of the last element in the list this is not good as in the caller side is done this:

`registrationData.UnstakedInfo = registrationData.UnstakedInfo[index:]`

This means that the last element which was already unBonded would be always kept. If the index returned will be greater than the last index in the list (equal with the len(registrationData.UnstakedInfo) ) then is ok

Reply:
Will add unit test :+1:

Reply:
Unit test showed that there was a bug there :1st_place_medal: